src/sql/services/notebook/notebookServiceImpl.ts

```diff
@@ -169,14 +181,25 @@ export class NotebookService extends Disposable implements INotebookService {
 			throw new Error(localize('notebookUriNotDefined', 'No URI was passed when creating a notebook manager'));
 		}
 		let uriString = uri.toString();
-		let manager = this._managers.get(uriString);
-		if (!manager) {
-			manager = await this.doWithProvider(providerId, (provider) => provider.getNotebookManager(uri));
-			if (manager) {
-				this._managers.set(uriString, manager);
+		let manager: INotebookManager[] = this._managers.get(uriString);
+		// If manager already exists for a given notebook, return it
+		if (manager) {
+			let index = manager.findIndex(m => m.providerId === providerId);
+			if (index && index >= 0) {
+				return manager[index];
 			}
 		}
-		return manager;
+		let newManager = await this.doWithProvider(providerId, (provider) => provider.getNotebookManager(uri));
+
+		// If no managers exist for a notebook, add this manager
+		// Otherwise, add manager to array
+		if (!manager) {
+			this._managers.set(uriString, [newManager]);
+		} else if (!manager.find(m => m.providerId === providerId)) {
+			manager.push(newManager);
+			this._managers.set(uriString, manager);
+		}
```

Comment:
How can this condition ever be true? didn't we search for the existing manager earlier and not find it? I think you need to fix this:
```suggestion
		managers = managers || [];
		managers.push[newManager];
		this._managers.set(uriString, managers);

```

Reply:
This set of code would replace the entire if/else block (trying out the suggested change preview but seems like it has issues).